src/soc/intel/baytrail/txei.c

```diff
@@ -0,0 +1,1196 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#define __SIMPLE_DEVICE__
+
+#include <bootmode.h>
+#include <console/console.h>
+#include <device/device.h>
+#include <device/pci.h>
+#include <device/pci_def.h>
+#include <device/pci_ops.h>
+#include <soc/iomap.h>
+#include <soc/pci_devs.h>
+#include <soc/pm.h>
+#include <soc/txe.h>
+#include <cf9_reset.h>
+#include <delay.h>
+#include <timer.h>
+
+#define SLOT_SIZE			sizeof(uint32_t)
+
+#define TXEI_HDR_IS_COMPLETE		(1 << 31)
+#define TXEI_HDR_LENGTH_START		16
+#define TXEI_HDR_LENGTH_SIZE		9
+#define TXEI_HDR_LENGTH			(((1 << TXEI_HDR_LENGTH_SIZE) - 1) << TXEI_HDR_LENGTH_START)
+#define TXEI_HDR_HOST_ADDR_START	8
+#define TXEI_HDR_HOST_ADDR		(((1 << 8) - 1) << TXEI_HDR_HOST_ADDR_START)
+#define TXEI_HDR_TXE_ADDR_START		0
+#define TXEI_HDR_TXE_ADDR		(((1 << 8) - 1) << TXEI_HDR_TXE_ADDR_START)
+
+enum txe_cmd_result {
+	TXE_CMD_RESULT_GLOBAL_RESET_REQUESTED,
+	TXE_CMD_RESULT_SUCCESS,
+	TXE_CMD_RESULT_ERROR,
+	TXE_CMD_RESULT_DISABLED,
+	TXE_CMD_RESULT_RETRY,
+};
+
+bool is_txe_devfn_visible(void)
+{
+	if (!is_devfn_enabled(PCI_DEVFN(TXE_DEV, TXE_FUNC))) {
+		printk(BIOS_WARNING, "TXE device 1a.0 is disabled\n");
+		return false;
+	}
+
+	if (pci_read_config16(PCI_DEV(0, TXE_DEV, TXE_FUNC), PCI_VENDOR_ID) == 0xFFFF) {
+		printk(BIOS_WARNING, "TXE device 1a.0 is hidden\n");
+		return false;
+	}
+
+	return true;
+}
+
+/* Get HECI BAR 0 from PCI configuration space */
+static uintptr_t get_txei_bar(pci_devfn_t dev)
+{
+	uintptr_t bar;
+
+	bar = pci_read_config32(dev, PCI_BASE_ADDRESS_0);
+	assert(bar != 0);
+	return bar & ~PCI_BASE_ADDRESS_MEM_ATTR_MASK;
+}
+
+static void txei_assign_resource(pci_devfn_t dev, uintptr_t tempbar0, uintptr_t tempbar1)
+{
+	u16 pcireg;
+
+	/* Assign Resources */
+	/* Clear BIT 1-2 of Command Register */
+	pcireg = pci_read_config16(dev, PCI_COMMAND);
+	pcireg &= ~(PCI_COMMAND_MASTER | PCI_COMMAND_MEMORY);
+	pci_write_config16(dev, PCI_COMMAND, pcireg);
+
+	/* Program Temporary BAR for TXEI device */
+	if (tempbar0 == 0)
+		pci_write_config32(dev, PCI_BASE_ADDRESS_0, TXEI_BASE_ADDRESS);
+	else
+		pci_write_config32(dev, PCI_BASE_ADDRESS_0, tempbar0);
+
+	if (tempbar1 == 0)
+		pci_write_config32(dev, PCI_BASE_ADDRESS_1, TXEI2_BASE_ADDRESS);
+	else
+		pci_write_config32(dev, PCI_BASE_ADDRESS_1, tempbar1);
+
+	/* Enable Bus Master and MMIO Space */
+	pci_or_config16(dev, PCI_COMMAND, PCI_COMMAND_MASTER | PCI_COMMAND_MEMORY);
+}
+
+int txe_aliveness_request(uintptr_t tempbar, bool request)
+{
+	struct stopwatch sw;
+
+	if (request)
+		write32p(tempbar + TXE_MMIO_ALIVENESS_REQ,
+			 read32p(tempbar + TXE_MMIO_ALIVENESS_REQ) | 1);
+	else
+		write32p(tempbar + TXE_MMIO_ALIVENESS_REQ,
+			 read32p(tempbar + TXE_MMIO_ALIVENESS_REQ) & (uint32_t)~1);
+
+	stopwatch_init_usecs_expire(&sw, HECI_READY_TIMEOUT);
+	while ((read32p(tempbar + TXE_MMIO_ALIVENESS_RESP) & request) != request) {
```

Comment:
If `request == false`, this statement will never be true.

```suggestion
	while ((read32p(tempbar + TXE_MMIO_ALIVENESS_RESP) & 1) != request) {
```

Unless that is the expected behavior, in which case it should be described in the comment above the function.